### PR TITLE
fix: fix calculating patch for respect ignore diff feature

### LINKED ISSUE #9678: sync is not working with ignoreDifferences and RespectIgnoreDifferences=true
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

I have an application where I want to ignore the image of the deployment and ignore it from being synced, I can see that the image is ignored and there is new env variables to be added in app diff but once I sync it it succeeds without applying the env variables like it's ignored with the image itself and I still see them in app diff!. This is the application:

```
project: default
source:
  repoURL: 'git@bitbucket.org:x/x.git'
  path: kubernetes/production
  targetRevision: master
  plugin:
    name: argocd-vault-plugin
destination:
  server: 'https://x.x.x.x'
syncPolicy:
  syncOptions:
    - RespectIgnoreDifferences=true
ignoreDifferences:
  - group: apps
    kind: Deployment
    name: service
    jqPathExpressions:
      - >-
        .spec.template.spec.containers[] | select(.name == "service").image
```

**To Reproduce**

Apply an application with ignoreDifferences like the application above and add new changes inside the same container specs
The new specs will not be applied after sync.

**Expected behavior**

I expect the image to be ignored and the new env variables to be applied/synced

**Version**

```shell
❯ argocd version --grpc-web
argocd: v2.2.7+b8e154f
  BuildDate: 2022-03-09T01:07:22Z
  GitCommit: b8e154f767412172bb0c2b41131460c34d2a0c73
  GitTreeState: clean
  GoVersion: go1.16.11
  Compiler: gc
  Platform: linux/amd64
argocd-server: v2.3.4+ac8b7df
  BuildDate: 2022-05-18T11:41:37Z
  GitCommit: ac8b7df9467ffcc0920b826c62c4b603a7bfed24
  GitTreeState: clean
  GoVersion: go1.17.10
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: v4.4.1 2021-11-11T23:36:27Z
  Helm Version: v3.8.0+gd141386
  Kubectl Version: v0.23.1
  Jsonnet Version: v0.18.0
```

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/9678 

When ignoreDifferences and `RespectIgnoreDifferences=true` are used the Argo CD needs to "copy" values of ignored paths from the k8s resource of the managed cluster to the expected resource. This way "ignored" fields won't be modified. The "copying" was implemented incorrectly for arrays. PR fixes the implementation.

It is still impossible to  ignoring differences in selected array elements. Created https://github.com/argoproj/argo-cd/issues/17694 to track fix.

### CODE REVIEW

**Comment by crenshaw-dev** [comment]
@alexmt I'm not sure how much you're trying to tackle in this PR, but [this issue](https://github.com/argoproj/argo-cd/issues/9678) has a lot of currently-broken use cases documented, and @blakepettersson converted many (all?) of them to [unit tests](https://github.com/argoproj/argo-cd/pull/14602/files#diff-3fc68b1ec0659b78121ecf80e444a7ea8092f5f514746c97664319f6c9cf7393) which maybe could be borrowed in this PR.

**Comment by alexmt** [comment]
@crenshaw-dev thank you for letting me know @crenshaw-dev . The PR solves the https://github.com/argoproj/argo-cd/issues/9678 partially: it is possible to ignore array completely but still impossible to ignore one element of the array. I propose to merge PR to unblock users who want to ignore array fields and continue working on improvement in https://github.com/argoproj/argo-cd/issues/17694

**Review by gdsoumya — APPROVED** [review]
LGTM

File: controller/sync_test.go [hunk 1]
```diff
@@ -455,3 +457,87 @@ func TestNormalizeTargetResources(t *testing.T) {
...
+				Group:             "projectcontour.io",
+				Kind:              "HTTPProxy",
+				JQPathExpressions: []string{".spec.routes[]"},
+				//JSONPointers: []string{"/spec/routes"},
+			},
+		}
+		f := setupHttpProxy(t, ignores)
+		target := test.YamlToUnstructured(testdata.TargetHTTPProxy)
+		y, _ := yaml.Marshal(target.Object)
+		fmt.Printf("\n--- target ---\n%s", string(y))
```
Comment (jessesuen) [inline controller/sync_test.go:498]: This is leftover from my debug staements. You may want to remove.

File: controller/sync_test.go [hunk 2]
```diff
@@ -455,3 +457,87 @@ func TestNormalizeTargetResources(t *testing.T) {
...
+	}
+
+	t.Run("will properly ignore nested fields within arrays", func(t *testing.T) {
+		// given
+		ignores := []v1alpha1.ResourceIgnoreDifferences{
+			{
+				Group:             "projectcontour.io",
+				Kind:              "HTTPProxy",
+				JQPathExpressions: []string{".spec.routes[]"},
+				//JSONPointers: []string{"/spec/routes"},
```
Comment (jessesuen) [inline controller/sync_test.go:492]: Is there any value in trying both jqpathexpression and jsonpointer? e.g. Should the test iterate the two options?

File: controller/sync_test.go (hunk 1 above)
Comment (alexmt) [reply controller/sync_test.go:498]: done! thank you

File: controller/sync_test.go [hunk 3]
```diff
@@ -455,3 +457,87 @@ func TestNormalizeTargetResources(t *testing.T) {
...
+		// assert some arbitrary field to show `entries[0]` is not an empty object
+		require.Equal(t, "sample-header", dig[string](f.comparisonResult.reconciliationResult.Live[0].Object, []interface{}{"spec", "routes", 0, "rateLimitPolicy", "global", "descriptors", 0, "entries", 0, "requestHeader", "headerName"}))
+
+		// target has 2 entries
+		require.Equal(t, 2, len(dig[[]any](f.comparisonResult.reconciliationResult.Target[0].Object, []interface{}{"spec", "routes", 0, "rateLimitPolicy", "global", "descriptors", 0, "entries"})))
+		// assert some arbitrary field to show `entries[0]` is not an empty object
+		require.Equal(t, "sample-header", dig[string](f.comparisonResult.reconciliationResult.Target[0].Object, []interface{}{"spec", "routes", 0, "rateLimitPolicy", "global", "descriptors", 0, "entries", 0, "requestHeaderValueMatch", "headers", 0, "name"}))
+
+		// It should be *2* entries in the array
+		require.Equal(t, 1, len(dig[[]any](patchedTargets[0].Object, []interface{}{"spec", "routes", 0, "rateLimitPolicy", "global", "descriptors"})))
```
[the remark this answers is not in the document]
Comment (alexmt) [reply controller/sync_test.go]: thank you. updated

**Comment by alexmt** [comment]
The e2e test was failing because my change introduced another bug. I had to implement strategic merge patch support and partially resolve https://github.com/argoproj/argo-cd/issues/17694 . So now user can ignore array elements with merge keys in well known k8s types. CRDs are not supported (same as in diffing).

I've added most of tests from https://github.com/argoproj/argo-cd/pull/14602 except Rollouts test (since it is CRD) and mutating webhook test (it is trying to merge arrays without merge key).

PTAL

**Comment by blakepettersson** [comment]
>I've added most of tests from https://github.com/argoproj/argo-cd/pull/14602 except mutating webhook test (it is trying to merge arrays without merge key).

I might be missing something here, but doesn't the mutating webhook test have merge keys in its arrays?

```yaml
webhooks:
  - admissionReviewVersions:
      - v1beta1
    clientConfig:
      caBundle: >-
        something-new
      service:
        name: aws-load-balancer-webhook-service
        namespace: kube-system
        path: /mutate-v1-pod
    failurePolicy: Fail
    name: mpod.elbv2.k8s.aws # <- merge key?
# Omitted other elements
```

Or do you mean that because each webhook has a `rules` array that does not have merge keys, strategic merge patching would not work?

File: controller/sync_test.go [hunk 4]
```diff
@@ -455,3 +455,207 @@ func TestNormalizeTargetResources(t *testing.T) {
...
+		assert.Equal(t, 1, len(ports))
+
+		env := containers[0].(map[string]interface{})["env"].([]interface{})
+		assert.Equal(t, 3, len(env))
+
+		first := env[0]
+		second := env[1]
+		third := env[2]
+
+		// Currently the defined order at this time is the insertion order of the target manifest.
```
Comment (blakepettersson) [inline controller/sync_test.go:558]: Is this still true when using SMP?

**Comment by alexmt** [comment]
@blakepettersson , thank you for review. I've clicked auto-merge and Github automatically merged PR once CI is green. Checking your comment about mutating webhook test

**Pull request merged**